File: packages/python/src/brainpick/brain_template.py
"""`brainpick init --template brain` (spec/85): scaffold a format-2 brain.

The files are the canonical tree integrations/brain-template/, shipped as package data
and synced by scripts/sync-brain-template.mjs; `{{today}}` and `{{bundle_id}}` are the
only placeholders. Nothing that exists is overwritten: an existing henxels.yaml gets
the brain contract as a fragment to paste instead.
"""
from __future__ import annotations

from pathlib import Path

TEMPLATES = ("brain",)
CONTRACT = "henxels.yaml"
FRAGMENT_MARKER = "  # --- the brainpick brain"
GITIGNORE_ENTRIES = ("_temp/", "brainpick.local.toml", ".brainpick/", "_todo.md")


def template_root() -> Path:
    """The shipped template: the package copy first (installed wheels), then the
    repo-root canonical (dev checkout)."""
    packaged = Path(__file__).resolve().parent / "_brain_template"
    if packaged.is_dir():
        return packaged
    return Path(__file__).resolve().parents[4] / "integrations" / "brain-template"


def template_files() -> list[str]:
    """Every file the template writes, as sorted bundle-relative POSIX paths."""
    base = template_root()
    return sorted(p.relative_to(base).as_posix() for p in base.rglob("*") if p.is_file())


def contract_fragment() -> str:
    """The brain rules of the template contract, to paste into an existing henxels.yaml."""
    text = (template_root() / CONTRACT).read_text(encoding="utf-8")
    return text[text.index(FRAGMENT_MARKER):]


def _ensure_gitignored(root: Path) -> str | None:
    """Append the missing entries to .gitignore: "created", "appended" or None."""
    gitignore = root / ".gitignore"
    existing = gitignore.read_text(encoding="utf-8") if gitignore.is_file() else None
    lines = {line.strip() for line in (existing or "").splitlines()}
    missing = [entry for entry in GITIGNORE_ENTRIES if entry not in lines and entry.rstrip("/") not in lines]
    if not missing:
        return None
    separator = "" if not existing or existing.endswith("\n") else "\n"
    with gitignore.open("w", encoding="utf-8", newline="") as handle:
        handle.write((existing or "") + separator + "".join(entry + "\n" for entry in missing))
    return "appended" if existing is not None else "created"


def scaffold_brain(root: Path, today: str, bundle_id: str) -> dict:
    """Write the template into `root`. Returns {"written", "existing"} (template paths),
    "gitignore" ("created" | "appended" | None) and "fragment" (the contract to paste
    when henxels.yaml already existed, else None)."""
    base = template_root()
    written: list[str] = []
    existing: list[str] = []
    for rel in template_files():
        target = root / rel
        if target.exists():
            existing.append(rel)
            continue
        text = (base / rel).read_text(encoding="utf-8")
        text = text.replace("{{today}}", today).replace("{{bundle_id}}", bundle_id)
        target.parent.mkdir(parents=True, exist_ok=True)
        with target.open("w", encoding="utf-8", newline="") as handle:
            handle.write(text)
        written.append(rel)
    return {
        "written": written,
        "existing": existing,
        "gitignore": _ensure_gitignored(root),
        "fragment": contract_fragment() if CONTRACT in existing else None,
    }

File: packages/node/src/brain-template.ts
/** `brainpick init --template brain` (spec/85): scaffold a format-2 brain. Twin of brain_template.py.
 *
 * The files are the canonical tree integrations/brain-template/, shipped in the npm package as
 * brain-template/ and synced by scripts/sync-brain-template.mjs; `{{today}}` and `{{bundle_id}}`
 * are the only placeholders. Nothing that exists is overwritten: an existing henxels.yaml gets
 * the brain contract as a fragment to paste instead.
 */
import { existsSync, mkdirSync, readdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";

import { cmpStr } from "./core/canonical";
import { pySplitLines, pyStrip } from "./core/pyfmt";
import { PACKAGE_ROOT } from "./version";

export const TEMPLATES = ["brain"] as const;
const CONTRACT = "henxels.yaml";
const FRAGMENT_MARKER = "  # --- the brainpick brain";
const GITIGNORE_ENTRIES = ["_temp/", "brainpick.local.toml", ".brainpick/", "_todo.md"];

export interface ScaffoldReport {
  written: string[];
  existing: string[];
  gitignore: "created" | "appended" | null;
  fragment: string | null;
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

/** The shipped template: the package copy first (installed tarballs), then the repo-root
 * canonical (dev checkout). */
export function templateRoot(): string {
  const packaged = resolve(PACKAGE_ROOT, "brain-template");
  if (isDirectory(packaged)) return packaged;
  return resolve(PACKAGE_ROOT, "..", "..", "integrations", "brain-template");
}

/** Every file the template writes, as sorted bundle-relative POSIX paths. */
export function templateFiles(): string[] {
  const files: string[] = [];
  const walk = (dir: string, prefix: string): void => {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const rel = prefix === "" ? entry.name : `${prefix}/${entry.name}`;
      if (entry.isDirectory()) walk(join(dir, entry.name), rel);
      else if (entry.isFile()) files.push(rel);
    }
  };
  walk(templateRoot(), "");
  return files.sort(cmpStr);
}

/** The brain rules of the template contract, to paste into an existing henxels.yaml. */
export function contractFragment(): string {
  const text = readFileSync(join(templateRoot(), CONTRACT), "utf8");
  return text.slice(text.indexOf(FRAGMENT_MARKER));
}

/** Append the missing entries to .gitignore: "created", "appended" or null. */
function ensureGitignored(root: string): "created" | "appended" | null {
  const path = join(root, ".gitignore");
  let existing: string | null = null;
  try {
    if (statSync(path).isFile()) existing = readFileSync(path, "utf8");
  } catch {
    existing = null;
  }
  const lines = new Set(pySplitLines(existing ?? "").map((line) => pyStrip(line)));
  const missing = GITIGNORE_ENTRIES.filter((entry) => !lines.has(entry) && !lines.has(entry.replace(/\/+$/, "")));
  if (missing.length === 0) return null;
  const separator = !existing || existing.endsWith("\n") ? "" : "\n";
  writeFileSync(path, (existing ?? "") + separator + missing.map((entry) => entry + "\n").join(""), "utf8");
  return existing !== null ? "appended" : "created";
}

/** Write the template into `root`: the template paths written and left as they were, the
 * .gitignore outcome, and the contract to paste when henxels.yaml already existed. */
export function scaffoldBrain(root: string, today: string, bundleId: string): ScaffoldReport {
  const base = templateRoot();
  const written: string[] = [];
  const existing: string[] = [];
  for (const rel of templateFiles()) {
    const target = join(root, rel);
    if (existsSync(target)) {
      existing.push(rel);
      continue;
    }
    const text = readFileSync(join(base, rel), "utf8").replaceAll("{{today}}", today).replaceAll("{{bundle_id}}", bundleId);
    mkdirSync(dirname(target), { recursive: true });
    writeFileSync(target, text, "utf8");
    written.push(rel);
  }
  return {
    written,
    existing,
    gitignore: ensureGitignored(root),
    fragment: existing.includes(CONTRACT) ? contractFragment() : null,
  };
}
